### PR TITLE
[JENKINS-26643] DescribableList.buildDependencyGraph should catch RuntimeException

### PR DESCRIPTION
**Pull request by recampbell**
[JENKINS-26643](https://issues.jenkins-ci.org/browse/JENKINS-26643)

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
It may make sense to handle <code>InterruptedException</code> and to exit the cycle.

**Comment by recampbell** [comment]
Thanks for the review.
@oleg-nenashev InterruptedException is a checked exception, how could I catch it?

I'm also supposed to request @reviewbybees

**Comment by oleg-nenashev** [comment]
* <code>InterruptedException</code> is being used to interrupt builds via Jenkins UI, timeouts, etc.
* API classes like <code>DescribableList</code> may be used from builds
* In the case of hanging extensions, the current handling may suppress the external abort event and call the next handler
* => A user may need to click the Cancel button several times

**Comment by jglick** [comment]
The current API does not permit `InterruptedException` (and AFAIK `buildDependencyGraph` is not called during builds) so I am not sure why that is even being discussed here.

File: core/src/main/java/hudson/util/DescribableList.java [hunk 1]
```diff
@@ -210,7 +212,12 @@ public void buildDependencyGraph(AbstractProject owner,DependencyGraph graph) {
         for (Object o : this) {
             if (o instanceof DependencyDeclarer) {
                 DependencyDeclarer dd = (DependencyDeclarer) o;
-                dd.buildDependencyGraph(owner,graph);
+                try {
+                    dd.buildDependencyGraph(owner,graph);
+                } catch (RuntimeException e) {
+                    String projectName = owner == null ? "unknown" : owner.getFullDisplayName();
+                    LOGGER.log(Level.SEVERE, "Failed to build dependency graph for " + projectName,e);
```
Comment (jglick) [inline core/src/main/java/hudson/util/DescribableList.java]: For log messages, better to use `getFullName`. Or simply

    LOGGER.log(Level.SEVERE, "Failed to build dependency graph for " + owner, e);

which will contain the same information (and more) anyway.

**Comment by oleg-nenashev** [comment]
@jglick 
I use DependencyGraph in several infrastructure backend jobs from Groovy. I'm not sure it's a valid *corner*-case BTW. I don't see such usage examples in plugins, so I agree the exception may be just logged.

> The current API does not permit InterruptedException

What do you mean (Javadoc/throws or something else)?

**Comment by jglick** [comment]
`DependencyDeclarer.buildDependencyGraph` does not declare any checked exceptions, so it is not possible for an implementation to throw `InterruptedException`.

**Comment by oleg-nenashev** [comment]
Whoops. In Jenkins it is not a runtime one. Please ignore my previous comments. LGTM

**Comment by stephenc** [comment]
:+1: though as a stylistic concern I prefer not to do string concatenation when logging messages, rather to construct the log record and set the parameters and throws (or use String.format) as that makes it easier to add additional details to the log message later... for concatenation of a fixed string and a variable string this is probably not an issue, but if you were doing somethling like 

```
LOGGER.log(Level.SEVERE, "Failed to build dependency graph for " + owner.getFullName() + "( " + owner.getFullDisplayName()+" ) : " + e.getMessage(),e)
```

Then I would insist on either

```
LogRecord lr = new LogRecord(Level.SEVERE, "Failed to build dependency graph for {0} ( {1} )");
lr.setThrown(e);
lr.setParameters(new Object[]{owner.getFullName(), owner.getDisplayName, e.getMessage()});
LOGGER.log(lr)
```

or 

```
LOGGER.log(Level.SEVERE, String.format("Failed to build dependency graph for %s ( %s ) : %s", owner.getFullName(), owner.getDisplayName, e.getMessage()),e);
```

The latter form is only relevant for log levels of WARNING or above, for log levels below WARNING you should use the LogRecord form as that prevents building the string if the log message is not being logged.

**Comment by recampbell** [comment]
Makes sense, thanks Stephen.

**Comment by jglick** [comment]
If you are logging at `INFO` or above, the message will almost certainly be printed, so you can just use whatever form is most convenient; string concatenation works fine, but `String.format` is sometimes more readable.

For `FINE` (or I guess `CONFIG`) or below, the normal idiom is to use the overloads taking format parameters directly:

    LOGGER.log(Level.FINE, "failed in {0}", something);

or

    LOGGER.log(Level.FINE, "failed in {0} because of {1}", new Object[] {something, somethingElse});

This does not work when a stack trace must be printed, so in the unusual case that you are displaying a stack trace but below `INFO`, you can use the more complicated `LogRecord` technique…or simply use plain string concatenation for the message but guard the whole thing with `isLoggable`:

    if (LOGGER.isLoggable(Level.FINE)) {
        LOGGER.log(Level.FINE, "failed in " + something, x);
    }

**Comment by stephenc** [comment]
Well I would argue that INFO is one level where users have a reasonable possibility of turning off... and haven been bitten by string concatenation causing excessive GC and CPU pressure in the past, my experience says WARNING or above is the only safe time to use string concatenation if you have a constructed message.

The `LOGGER.isLoggable` assumes that the logger that will log that issue actually cares about the message string (might just be counting log noise) so I argue you should always let LOGGER decide what to do... but if you are doing anything expensive in stringifying your arguments, or if the string form of the argument depends on mutable state then wrap that also.

You only need the LogRecord pattern for cases like Ryan's where there is a compound message **and** a stack trace... obviously if you have one or the other you are safe to use the standard overloads.

Finally, if the string form of the state you are logging depends on mutable state then you need to explicitly call the toString rather than leaving it up to the LOGGER and that will need wrapping in isLoggable.

With WARNING and above, they should be rare enough that string concatenation (or the safer and easier to see the text of `String.format`) is sufficient

**Comment by jglick** [comment]
All true, though these are performance optimizations for very unlikely events.

**Pull request merged**